ugc/types.hpp

```diff
@@ -199,6 +243,84 @@ struct Attribute
   TranslationKey m_value{};
 };
 
+struct ReviewFeedback
+{
+  ReviewFeedback() = default;
+  ReviewFeedback(Sentiment const sentiment, Time const & time)
+    : m_sentiment(sentiment), m_time(time)
+  {
+  }
+
+  Sentiment m_sentiment{};
+  Time m_time{};
+};
+
+struct ReviewAbuse
+{
+  ReviewAbuse() = default;
+  ReviewAbuse(std::string const & reason, Time const & time) : m_reason(reason), m_time(time) {}
+
+  std::string m_reason{};
+  Time m_time{};
+};
+
+namespace v0
+{
+struct UGCUpdate
+{
+  UGCUpdate() = default;
+  UGCUpdate(Ratings const & records, Text const & text, Time const & time)
+    : m_ratings(records), m_text(text), m_time(time)
+  {
+  }
+
+  DECLARE_VISITOR(visitor(m_ratings, "ratings"), visitor(m_text, "text"), visitor(m_time, "date"))
+
+  bool operator==(UGCUpdate const & rhs) const
+  {
+    return m_ratings == rhs.m_ratings && m_text == rhs.m_text && m_time == rhs.m_time;
+  }
+
+  bool IsEmpty() const
+  {
+    return !((!m_ratings.empty() || !m_text.m_text.empty()) && m_time != Time());
```

Comment:
I guess you may simplify the expression. Too many parens.

Reply:
ok